src/web/components/ClientSideTokenGeneration/CstgEditDialog.tsx

```diff
@@ -0,0 +1,101 @@
+import { FormProvider, useForm } from 'react-hook-form';
+
+import { Dialog } from '../Core/Dialog';
+import { RootFormErrors } from '../Input/FormError';
+import { TextInput } from '../Input/TextInput';
+import {
+  CstgValueType,
+  EditCstgValuesFormProps,
+  extractTopLevelDomain,
+  validateAppId,
+} from './CstgHelper';
+
+type CstgEditDialogProps = Readonly<{
+  cstgValue: string;
+  existingCstgValues: string[];
+  onEdit: (newCstgValue: string, originalCstgValue: string) => Promise<boolean>;
+  onOpenChange: () => void;
+  cstgValueType: CstgValueType;
+}>;
+
+function CstgEditDialog({
+  cstgValue,
+  onEdit,
+  onOpenChange,
+  existingCstgValues,
+  cstgValueType,
+}: CstgEditDialogProps) {
+  const formMethods = useForm<EditCstgValuesFormProps>({
+    defaultValues: {
+      cstgValue,
+    },
+  });
+
+  const {
+    handleSubmit,
+    setError,
+    formState: { errors },
+  } = formMethods;
+
+  const showInvalidError = () => {
+    setError('root.serverError', {
+      type: '400',
+      message: `Edited value is an invalid ${cstgValueType.toLowerCase()}.`,
+    });
+  };
+
+  const onSubmit = async (formData: EditCstgValuesFormProps) => {
+    let updatedCstgValue = formData.cstgValue;
+    const originalCstgValue = cstgValue;
+
+    if (cstgValueType === CstgValueType.Domain) {
+      updatedCstgValue = extractTopLevelDomain(updatedCstgValue);
+    }
+
+    if (updatedCstgValue === originalCstgValue) {
+      onOpenChange();
+    } else if (existingCstgValues.includes(updatedCstgValue)) {
+      setError('root.serverError', {
+        type: '400',
+        message: `${cstgValueType} already exists.`,
+      });
+    } else if (cstgValueType === CstgValueType.MobileAppId) {
+      if (!validateAppId(updatedCstgValue)) {
+        showInvalidError();
+      }
+    } else {
+      const editSuccess = await onEdit(updatedCstgValue, originalCstgValue);
+      if (!editSuccess) {
+        showInvalidError();
+      }
+    }
```

Comment:
Again here, may look a little weird because validating an edited app id = front end, validating an edited top level domain = back end

Reply:
Perhaps injecting an optional validator (and maybe an optional transformer function as well - e.g. for domains it would be `extractTopLevelDomain`) might have been cleaner, but I think this approach is OK in this case. I don't think we're likely to extend it to 3+ types of things this table handles, and we can always tweak it later.